src/Ilios/CoreBundle/Classes/CourseRollover.php

```diff
@@ -0,0 +1,477 @@
+     * @param ManagerInterface $offeringManager
+     * @param ManagerInterface $objectiveManager
+     */
+    public function __construct(
+        ManagerInterface $courseManager,
+        ManagerInterface $learningMaterialManager,
+        ManagerInterface $courseLearningMaterialManager,
+        ManagerInterface $sessionManager,
+        ManagerInterface $sessionLearningMaterialManager,
+        ManagerInterface $offeringManager,
+        ManagerInterface $objectiveManager
+    ) {
+        $this->courseManager = $courseManager;
+        $this->learningMaterialManager = $learningMaterialManager;
+        $this->courseLearningMaterialManager = $courseLearningMaterialManager;
+        $this->sessionManager = $sessionManager;
+        $this->sessionLearningMaterialManager = $sessionLearningMaterialManager;
+        $this->offeringManager = $offeringManager;
+        $this->objectiveManager = $objectiveManager;
+
+    }
+
+    /**
+     * @param int $courseId
+     * @param int $newAcademicYear
+     * @param array $options
+     * @return CourseInterface the new, rolled-over course.
+     * @throws \Exception
+     */
+    public function rolloverCourse($courseId, $newAcademicYear, $options)
+    {
+        //now, get/set the required values from the provided arguments
+        $originalCourseId = $courseId;
+        $newStartDate = (!empty($options['new-start-date'])) ? new \DateTime($options['new-start-date']) : null;
+
+        //make sure that the new course's academic year or new start date year is not in the past
+        $this->confirmYearIsNotInPast($newAcademicYear);
+        if (!empty($newStartDate)) {
+            $this->confirmYearIsNotInPast($newStartDate->format('Y'));
+        }
+
+        //get the original course object
+        $originalCourse = $this->getOriginalCourse($originalCourseId);
+
+        //if a new title is to be used, update before checking for duplicates
+        $newTitle = !empty($options['new-course-title']) ? $options['new-course-title'] : $originalCourse->getTitle();
+        //before creating the newCourse object, check for courses with same title & year, so a rollover is not run 2x
+        $this->checkForDuplicateRollover($newTitle, $newAcademicYear);
+
+        //get/set the original course's start/end dates for use in calculation of offset
+        $originalCourseStartDate = $originalCourse->getStartDate();
+        $originalCourseEndDate = $originalCourse->getEndDate();
+
+        //get/set the offset in weeks and its +/- sign and set it as globally-accessible string in the class
+        $weeksOffset = $this->calculateRolloverOffsetInWeeks($originalCourse, $newAcademicYear, $newStartDate);
+
+        //taken with gratitude from http://php.net/manual/en/function.gmp-sign.php#106246
+        $weeksOffsetModifier = ( $weeksOffset > 0 ) ? 1 : ( ( $weeksOffset < 0 ) ? -1 : 0 );
+        $offsetInWeeks = (($weeksOffsetModifier < 0) ? '-' : '+') . ' ' . abs($weeksOffset) . ' weeks';
```

Comment:
is it possible to end up with a `0` value for `$weekOffsetModifier`? if so, then the code should check for that and skip the start/end-date modification below.